src/rgw/services/svc_notify.cc

```diff
@@ -394,96 +394,41 @@ int RGWSI_Notify::robust_notify(const DoutPrefixProvider *dpp,
 {
   // The reply of every machine that acks goes in here.
   boost::container::flat_set<std::pair<uint64_t, uint64_t>> acks;
-  bufferlist bl, rbl;;
+  bufferlist bl;
   encode(cni, bl);
 
   // First, try to send, without being fancy about it.
-  auto r = notify_obj.notify(dpp, bl, 0, &rbl, y);
+  auto r = notify_obj.notify(dpp, bl, 0, nullptr, y);
 
-  // If that doesn't work, get serious.
   if (r < 0) {
-    ldpp_dout(dpp, 1) << "robust_notify: If at first you don't succeed: "
+    ldpp_dout(dpp, 1) << __PRETTY_FUNCTION__ << ":" << __LINE__
+		      << " Notify failed: "
 		      << cpp_strerror(-r) << dendl;
+    ClibBackTrace bt(1);
+    ldpp_dout(dpp, 1) << __PRETTY_FUNCTION__ << ":" << __LINE__
+		      << " Backtrace of the operation where this happened: "
+		      << bt << dendl;
+  }
 
-
-    auto p = rbl.cbegin();
-    // Gather up the replies to the first attempt.
-    try {
-      uint32_t num_acks;
-      decode(num_acks, p);
-      // Doing this ourselves since we don't care about the payload;
-      for (auto i = 0u; i < num_acks; ++i) {
-	std::pair<uint64_t, uint64_t> id;
-	decode(id, p);
-	acks.insert(id);
-	ldpp_dout(dpp, 20) << "robust_notify: acked by " << id << dendl;
-	uint32_t blen;
-	decode(blen, p);
-	p += blen;
-      }
-    } catch (const buffer::error& e) {
-      ldpp_dout(dpp, 0) << "robust_notify: notify response parse failed: "
-		    << e.what() << dendl;
-      acks.clear(); // Throw away junk on failed parse.
-    }
-
-
-    // Every machine that fails to reply and hasn't acked a previous
-    // attempt goes in here.
-    boost::container::flat_set<std::pair<uint64_t, uint64_t>> timeouts;
-
-    auto tries = 1u;
-    while (r < 0 && tries < max_notify_retries) {
-      ++tries;
-      rbl.clear();
-      // Reset the timeouts, we're only concerned with new ones.
-      timeouts.clear();
-      r = notify_obj.notify(dpp, bl, 0, &rbl, y);
+  // If we timed out, get serious.
+  if (r == -ETIMEDOUT) {
+    RGWCacheNotifyInfo info;
+    info.op = REMOVE_OBJ;
```

Comment:
so REMOVE_OBJ removes an object from the cache?  or is it pretending that we removed the thing for real, but having the same effect?

Reply:
The latter. ObjectCache::remove marks the cache entry as invalid, removes it from the LRU, and erases it from the map.

It might be better to have an explicit INVALIDATE message for clarity/futureproofing. (And making the object lose its place in the LRU might be bad.)

The only downside would be an older RGW in the same cluster not knowing the new notify type. Is it unsupported for users to do that? (At least outside the context of an upgrade where they're all GOING to be upgraded?)

Reply:
now that we have caching of negative entries (cc @ofriedma), the risk would be that the receiving rgw might handle `REMOVE_OBJ` by creating a negative entry instead of just invalidating. that's not what we currently do - to create a negative entry, you have to call `ObjectCache::put()` with a `info.status = ENOENT` - but it would seem reasonable for a `REMOVE_OBJ` message to do that

to avoid that confusion, maybe just rename the existing `REMOVE_OBJ` to `INVALIDATE`? that would clarify the intent without backward compatibility issues

Reply:
+1

Reply:
Sounds good to me.